### DISCUSSION TITLE
Default posh-git prompt

### DISCUSSION THREAD

**Discussion by daniporr** [JanDeDobbeleer/oh-my-posh#2773] [Troubleshoot] [answered]
Hi, I would like to get the [default posh-git prompt](https://github.com/dahlbyk/posh-git#git-status-summary-information).
Is there a property that allows to do that?

P.S. I do not want to construct if from scratch since it is already provided by posh-git.

**Comment by JanDeDobbeleer** [comment, marked as answer]
@daniporr the exact same prompt? Because I just refactored everything so that posh-git and git are perfectly aligned. You don't really need to do anything other than enable posh-git, the default rendering is already providing all information correctly.

**Reply by daniporr** [reply]
@JanDeDobbeleer thank you for your very quick answer.
I tried the default template but, since I do not use a Nerd Font, it doesn't work properly. Therefore, I wanted to try the default one of posh-git which seems to use only characters that should be visible in PowerShell.

Moreover, I am running it on a quite slow machine, so I was thinking that having that one would be faster (but I am not sure about this).

After reading your message, I am now wondering if I can simply try to adapt your default template simply changing all those custom Unicode symbols.
Do you think that would be straightforward?

**Reply by JanDeDobbeleer** [reply]
@daniporr you can have a look at the minimal [themes](https://ohmyposh.dev/docs/themes), they only use standard icons.

**Reply by daniporr** [reply]
@JanDeDobbeleer thank you for your answer.
Let me clarify that I am using a "custom" theme derived from one of yours.

But, building on your answer, I tried to copy in my custom theme the git segments that you can find in the minimal themes. I tried all the minimal themes, but for almost all of them I still see some "boxes" (unrecognised characters).
The only ones which seem to not require additional characters are "onehalf.minimal" and "probua.minimal". Unfortunately, they are extremely different from the default one of posh-git (both in "shape" and in "colors").

Moreover, it seems that we are just re-writing from scratch (with a slightly different syntax) what posh-git already does. This was what I mentioned in my first message that wanted to avoid for compatibility reasons (if they update the posh-git command line, I want to automatically get those changes) and for efficiency reasons (posh-git already does the computations needed to draw that line, so there is no need to duplicate that).

Do you think it could be useful to have a property for the template that simply draws the default posh-git prompt? (in order to get all the benefits mentioned above)

**Reply by JanDeDobbeleer** [reply]
Actually, posh-git doesn't already do this unless we tell it to. Meaning we hijack the prompt and call posh-git, but only to get the raw status (so not pretty rendered). We do this to provide auto completion, and then provide it to oh-my-posh to avoid fetching all relevant git information twice. Initially we did make posh-git render the raw string, and displayed that but then you don't have all the functionality the git segment offers like colors based on git context or templating capabilities. So the `efficiency reasons` statement is incorrect. If you want to achieve this, hook into oh-my-posh like this:

````powershell
function Set-PoshGitStatus {
    $global:GitStatus = Get-GitStatus
    $env:POSH_GIT_STRING = Write-GitStatus -Status $global:GitStatus
}
New-Alias -Name 'Set-PoshContext' -Value 'Set-PoshGitStatus' -Scope Global -Force
```` 

You can then use this in a text segment leveraging environment variables:

```json
"template": "{{ if .Env.POSH_GIT_STRING }} {{ .Env.POSH_GIT_STRING }} {{ end }}"
```
You might to tweak the `Set-PoshGitStatus` to see if it is in a repository, I don't know how posh-git displays that.

**Reply by daniporr** [reply]
@JanDeDobbeleer thank you for your very detailed answer

> Actually, posh-git doesn't already do this unless we tell it to. Meaning we hijack the prompt and call posh-git, but only to get the raw status (so not pretty rendered). We do this to provide auto completion, and then provide it to oh-my-posh to avoid fetching all relevant git information twice. Initially we did make posh-git render the raw string, and displayed that but then you don't have all the functionality the git segment offers like colors based on git context or templating capabilities. So the `efficiency reasons` statement is incorrect.

I did not know that (and did not know it was actually possible).
I apologise for the wrong statement.
This seems a very clever way of using the functionalities provided by posh-git!

> If you want to achieve this, hook into oh-my-posh like this:

I will try your suggestion in the next few days
Thank you for that!

---

I think this question is now answered, but I am still curious about something.
Do you know why for several minimal themes I see some "boxes" (unrecognised characters) when I am in a folder containing a Git repository (i.e., when I use the git segment)?

**Reply by JanDeDobbeleer** [reply]
@daniporr yes, that's because the font you use doesn't have those icons. We make use of [nerd fonts](https://www.nerdfonts.com/), which you can also easily install using `oh-my-posh font install` (reload the terminal afterwards).

**Reply by daniporr** [reply]
@JanDeDobbeleer I apologise, probably my question was not very clear.
Let me clarify it.

My previous statement:

> several minimal themes I see some "boxes" (unrecognised characters) when I am in a folder containing a Git repository (i.e., when I use the git segment)

Your answer:

> yes, that's because the font you use doesn't have those icons. We make use of [nerd fonts](https://www.nerdfonts.com/), which you can also easily install using oh-my-posh font install (reload the terminal afterwards).

But, in the themes page there is written:

> Themes with `minimal` in their names do not require a Nerd Font. Read about [🆎Fonts](https://ohmyposh.dev/docs/installation/fonts) for more information.

And you also mentioned this in a previous comment:

> @daniporr you can have a look at the minimal [themes](https://ohmyposh.dev/docs/themes), they only use standard icons.

Therefore, the question is: *why do `minimal` themes require Nerd Font for the git segment if they are supposed to not require it at all (i.e., to only use standard icons)?*

P.S. if you prefer, I can open a different discussion/issue about this.

**Reply by JanDeDobbeleer** [reply]
@daniporr good question, I think the minimal fonts were created to support Cascadia Code, which ships out of the box with Windows Terminal.

**Reply by daniporr** [reply]
> @daniporr good question, I think the minimal fonts were created to support Cascadia Code, which ships out of the box with Windows Terminal.

I think you meant "minimal themes" not _"minimal fonts"_ (but let me know if this is not the case).

I use Cascadia Code, but those symbols are not available there (see https://github.com/microsoft/cascadia-code).
For instance, the branch symbol:

![image](https://user-images.githubusercontent.com/90617254/190143307-09603ebf-2eef-482b-8906-61470d50a04e.png)

Have I misunderstood something?

**Reply by JanDeDobbeleer** [reply]
Aha, but that should've been removed right? I didn't create the minimal themes so I rely on contributors to get that right. A prerequisite of oh-my-posh has always been nerd fonts. If indeed Cascadia Code doesn't have the branch icon, it should have an override in the [minimal configuration](https://github.com/JanDeDobbeleer/oh-my-posh/blob/620187aada8ab142753a0f64ab574c6f0631a86e/themes/agnoster.minimal.omp.json#L41).

**Reply by daniporr** [reply]
Yes, I agree! 😃
(that was exactly my point)

Thank you very much for your support and the useful conversation.

**Comment by daniporr** [comment]
### A custom template to mimic the posh-git prompt

I agree that the best answer to this discussion is:
https://github.com/JanDeDobbeleer/oh-my-posh/discussions/2773#discussioncomment-3635809

But I wanted to see how difficult it would be to faithfully mimic the posh-git prompt using the oh-my-posh properties, so I have created the following segment:

```json
{
  "foreground": "cyan",
  "foreground_templates": [
    "{{ if and (gt .Ahead 0) (gt .Behind 0) }}yellow{{ end }}",
    "{{ if gt .Ahead 0 }}green{{ end }}",
    "{{ if gt .Behind 0 }}red{{ end }}"
  ],
  "properties": {
    "branch_icon": "",
    "commit_icon": ""
  },
  "style": "diamond",
  "leading_diamond": "<lightYellow>[</>",
  "trailing_diamond": "<lightYellow>]</>",
  "template": "{{ .HEAD }}{{if .BranchStatus }} {{ .BranchStatus }}{{ end }}{{ if .Staging.Changed }} <green>{{ .Staging.String }}</>{{ end }}{{ if and (.Working.Changed) (.Staging.Changed) }} |{{ end }}<red>{{ if .Working.Changed }} {{ .Working.String }}{{ end }}{{ if and (.Working.Changed) (not .Staging.Changed) }} !{{ end }}{{ if and (.Working.Changed) (.Staging.Changed) }} ~{{ end }}</>",
  "type": "git"
}
```

Following the description in the official repository as faithfully as possible:
https://github.com/dahlbyk/posh-git#git-status-summary-information

I hope this can be useful to someone.

Please let me know if you think it can be improved (and how you think it can be improved)
